src/v037/server.rs

```diff
@@ -19,20 +23,61 @@ use tendermint::v0_37::abci::{
     MempoolResponse, Request, Response, SnapshotRequest, SnapshotResponse,
 };
 
+enum SocketKind {
+    Tcp(TcpStream),
+    Uds(UnixStream),
+}
+
+impl SocketKind {
+    fn split<'a>(
+        &'a mut self,
+    ) -> (
+        Box<dyn 'a + AsyncRead + Send + Unpin>,
+        Box<dyn 'a + AsyncWrite + Send + Unpin>,
+    ) {
+        match self {
+            Self::Tcp(tcp) => {
+                let (read, write) = tcp.split();
+                (Box::new(read), Box::new(write))
+            }
+            Self::Uds(uds) => {
+                let (read, write) = uds.split();
+                (Box::new(read), Box::new(write))
+            }
```

Comment:
I don't think this is actually necessary, callers have to match on the `SocketKind` anyway. We can just extract the inner value and leave the rest of the code unchanged (`socket.split()`).